server/eclipse-project/src/main/java/au/org/democracydevelopers/corla/model/vote/IRVChoices.java

```diff
@@ -21,5 +21,234 @@
+ * Main methods:
+ * - IsValid(): checks whether the choices are a valid IRV vote (without skipped or repeated
+ * preferences, or repeated candidates).
+ * - ValidInterpretation(): applies Colorado's IRV rules and returns the valid interpretation, as
+ * an ordered list of candidate names with the highest preference first.
+ */
 public class IRVChoices {
+
+  // Always sorted by preference, from highest preference (i.e. lowest number) to lowest
+  // preference (i.e. highest number). Repeated or skipped preferences are allowed.
+  private List<IRVPreference> choices;
+
+  public int getLength() {
+    return choices.size();
+  }
+
+  /**
+   * Constructor - takes a list of IRVPreferences.
+   * @param mutableChoices a list of IRVPreferences, which need not be valid - repeats or skipped
+   *                       preferences are allowed.
+   */
+  public IRVChoices(List<IRVPreference> mutableChoices) {
+    mutableChoices.sort(IRVPreference::compareTo);
+    choices = Collections.unmodifiableList(mutableChoices);
+  }
+
+  /**
+   * Constructor - takes IRV preferences as a string of the kind stored in the corla database,
+   * of the form "name1(p1), name2(p2), ..."
+   * Parses each individual preference into an IRVPreference and then calls the previous
+   * constructor.
+   * @param sanitizedChoices the IRV preferences as a string, which need not be a valid IRV vote -
+   *                         repeats or skipped preferences are allowed.
+   */
+  public IRVChoices(String sanitizedChoices) throws IRVParsingException {
+    this(parseChoices(sanitizedChoices));
+  }
+
+  /** Check whether the ballot is a valid list of preferences without repeats of candidate names
+   * or ranks and without skipped or invalid preferences.
+   */
+  public boolean IsValid() {
+
+    // Test for a perfect integer sequence of length choices.size(), by filling in the i-th element
+    // when i is read. preferencesInOrder[i-1] is true if we've encountered preference i - if
+    // we encounter it again, or encounter an out-of-bounds preference, return false.
+    boolean[] preferencesInOrder = new boolean[choices.size()];
+    for(IRVPreference p : choices) {
+      // If this preference is out of bounds or already used, the preference list is invalid
+      if (p.rank < 1 || p.rank > choices.size() || preferencesInOrder[p.rank-1]) {
+        return false;
+      }
+      // Otherwise, it's fine so far. Remember that we've seen this preference.
+      preferencesInOrder[p.rank-1] = true;
+    }
+
+    // Now check for repeat candidate names.
+    Set<String> candidateNamesSet = new HashSet<>();
+    List<String> candidateNames = choices.stream().map(c -> c.candidateName).collect(Collectors.toList());
+
+    // Set::add returns false if the item is already present, in which case this returns false.
+    // (Repeated candidate names are invalid.)
+    return candidateNames.stream().map(candidateNamesSet::add).reduce(true, Boolean::logicalAnd);
+  }
+
+  /**
```

Comment:
Good point - I've given them their full reference - 26.7.x - and a descriptor.

Also good point about what happens when it's called on valid votes - and yes it will be, because that'll be how we translate them from Dominion-style to ordered-list style.